analysea/tide.py

```diff
@@ -113,86 +102,107 @@ def demean_amps_phases(
     return const, mean_amps, mean_phases  # ignore mypy
 
 
-def calc_constituents(ts: pd.Series, kwargs: Dict[str, Any]) -> Any:
+def calc_constituents(
+    ts: pd.Series, resample_time: int = 10, lat: float = 0.0, **kwargs: Optional[Dict[str, Any]]
```

Comment:
type hinting for args and kwargs is a PITA...

It's only last week that I have been reading about this for the Nth time and I realized that I was doing it wrong.

For starters, `Optional` is not needed in this case. `arg: Optional[dict[str, Any]]`  means that you can call the function with either `arg=dict(a=1, b="2")` or `arg=None` which is not what you intend to do here.

Since you are using `**kwargs` it means that you may pass any number of keyword arguments and these (if any) will be available in the function scope in the dictionary named `kwargs`.

Now how do you typeset `kwargs`? Well there is no good answer. 

If you do know the names of the potential keyword arguments, then you can use `TypedDict` and `Unpack`. To understand this run `mypy --strict` on the following:

``` python
from typing_extensions import Unpack
from typing import TypedDict

class Movie(TypedDict):
    name: str
    year: int

# This function expects two keyword arguments - `name` of type `str`
# and `year` of type `int`. In this case, both `name` and `year` are mandatory
# but it is possible to mark them as optional, too. For more info check the docs: https://docs.python.org/3/library/typing.html#typing.TypedDict
def foo(**kwargs: Unpack[Movie]) -> None:
    print(kwargs)

foo()  # mypy error
foo(name="Name")  # mypy error
foo(name="Name", year=2023)  # correct
```

Now, if you don't know the names of the keyword arguments then there are two cases:

1. If you do know that the type of **all** the keyword arguments are of the _same_ type, then use this type, e.g. `**kwargs: str` or `**kwargs: int`. 
2. If you don't know the type of the keyword arguments, then you can use either `**kwargs: Any` or `**kwargs: object`. For a discussion between the two check here: https://stackoverflow.com/questions/39817081/typing-any-vs-object I think that for wrappers around 3rd party libs, we should be using the `Any` (but I haven't really thought about it).

For a more in length discussion which also covers typesetting `*args` check here: https://stackoverflow.com/a/73738408/592289

Reply:
will fix this later

Reply:
implemented  
`Unpack[UTideArgs]`

with: 
```
class UTideArgs(TypedDict):
    constit: NotRequired[str]
    method: NotRequired[str]
    order_constit: NotRequired[str]
    Rayleigh_min: NotRequired[float]
    lat: str
    verbose: NotRequired[str]
```

tailored for utide's `kwargs`